design/one-pager-crd-scaling.md

```diff
@@ -0,0 +1,590 @@
+  `_numLevels * _countersPerLevel * (Int64 + Uint64) = 7 * 4096 * (8+8) = 448
+  KB`! Thus, for 780 CRDs (plus other `zap` loggers), we are using `815 * 448 KB
+  = ~356 MB` of heap space. `zapcore.newCounters` is responsible for the ~31% of
+  the heap space allocated! Some deeper investigation reveals that
+  `apiextensions-apiserver` uses [`apiserver.crdHandler`] to serve the `/apis`
+  endpoint, and `apiserver.crdHandler` in turns embodies a
+  [`customresource.CustomResourceStorage`] for each served version of a CRD.
+  `customresource.CustomResourceStorage` represents the etcd v3 storage, which
+  in turn embodies an etcd v3 client and that client [has][etcdv3-client-logger]
+  a `zap.Logger`. These relationships reveal that we spend 448 KB of heap space
+  for each version served of each CRD. That's currently rougly 5 times larger
+  than the total amount of heap space needed to store the deserialized OpenAPI
+  v3 schemas of `provider-jet-aws` CRDs.
+- We allocate persistent (during the lifetime of an installed CRD) heap space
+  that's proportional to the size of the OpenAPI v3 schema of a CRD in the
+  [`SharedIndexInformer`] caches of various CRD controllers such as the
+  [`establish.EstablishingController`] or the [`finalizer.CRDFinalizer`]
+  controller, and for some other purposes such as CRD resource handler caches.
+  Currently, for `provider-jet-aws@0.4.0-preview`, such caches consume less than
+  ~80 MB of heap space.
+- However, to exacerbate the situation, we have generated ~190 CRDs each with
+  5000 OpenAPI v3 schema properties. Even with only 190 such CRDs, heap
+  profiling [data][kube-apiserver-heap-profile-5000-props] reveals that ~5 GB of
+  heap space is allocated and `zapcore.newCounters` is no longer the largest
+  allocation site anymore, as expected. This profiling data better emphasizes
+  allocation sites which are inflated with large (deserialized) OpenAPI v3 CRD
+  schemas. One can generate such CRDs using the following command:
+```
+github.com/ulucinar/terrajet-scale/cmd/apigen generate -g 200 -c 200 --property-count 5000
+```
+
+
+## Criteria Set for Ideal State
+
+In this section, we would like to discuss some Crossplane scenarios that we
+would like to support and that involve large numbers of CRDs:
+1. Installation of a single provider with hundreds of CRDs: As discussed above,
+   installation of a single provider, such as `provider-jet-aws` preview
+   edition, results in ~190 GVs being served by the API server. With the current
+   (`v1.23`) set of client-side throttling parameters of `kubectl`, this has
+   adverse effects on the perceived user performance. However, situation
+   improves as the burstiness and the fill rate of the token bucket rate limiter
+   are increased in `kubectl@v1.24`.
+2. Installation of multiple providers, such as `provider-jet-{aws,gcp,azure}`
+   preview editions, into the same cluster results in ~370 GVs being served by
```

Comment:
As discussed in the `Client-side Throttling` section, experiments done with a custom build of `kubectl` that allows us to specify discovery-client's tbrl parameters reveal that, with tbrl(b=400, r=50.0 qps), client-side throttling is no longer a bottleneck for our current GV counts (despite a delay of ~18s. on a cold cache). This is subject to change if the number of GVs increase because of different combinations of providers or because of API regroupings etc. However, the upstream community sees bumping these limits to cover certain use-cases as increasing the debt ceiling.